### PR TITLE
Add Grid tiler

### LINKED ISSUE #99: Add Grid Tiler
We need a Tiler to extract the tiles from a "grid". Tiles may be overlapping or not, depending on user choice.

### PR DESCRIPTION
**Pull request by alessiamarcolini**
Fix #99

### CODE REVIEW

File: src/histolab/tiler.py [hunk 1]
```diff
@@ -1,21 +1,250 @@
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 from typing import Tuple
 
 import numpy as np
 import sparse
+from typing_extensions import Protocol, runtime_checkable
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:6]: considering that 3.7 already have this 2 module in the typing std lib ... here I would use a try except statement to import them from the standard lib and if exception is raised let's import from typing_extension

File: tests/unit/test_tiler.py [hunk 2]
```diff
@@ -323,3 +352,283 @@ def tile_filename_fixture(self, request):
...
+            ),
+            (
+                CoordinatePair(0, 10, 0, 10),
+                CoordinatePair(0, 10, 0, 10),
+                True,
+                [True, False],
+                1,
+            ),
+        ),
+    )
```
Comment (ernestoarbitrio) [inline tests/unit/test_tiler.py:527]: I would avoid using 100 lines fixture in the parametrize decorator, I do prefer to have a separate fixture method to be passed as arg in the test function

File: src/histolab/tiler.py [hunk 3]
```diff
@@ -1,21 +1,250 @@
...
+
+        return tile_filename
+
     @abstractmethod
     def extract(self, slide: Slide):
         raise NotImplementedError
 
 
+class GridTiler(Tiler):
+    def __init__(
```
Comment (nicolebussola) [inline src/histolab/tiler.py:105]: docstring?

File: src/histolab/tiler.py [hunk 4]
```diff
@@ -1,21 +1,250 @@
...
+                    y_br_lvl = y_ul_lvl + tile_h_lvl
+
+                    tile_wsi_coords = scale_coordinates(
+                        reference_coords=CoordinatePair(
+                            x_ul_lvl, y_ul_lvl, x_br_lvl, y_br_lvl
+                        ),
+                        reference_size=slide.level_dimensions(level=self.level),
+                        target_size=slide.level_dimensions(level=0),
+                    )
+                    yield tile_wsi_coords
```
[the remark this answers is not in the document]
Comment (alessiamarcolini) [reply src/histolab/tiler.py:230]: working on it :D

File: src/histolab/tiler.py [hunk 5]
```diff
@@ -7,21 +7,317 @@
...
+            Iterator of tiles' CoordinatePair
+        """
+        box_mask_lvl = self.box_mask_lvl(slide)
+
+        regions = regions_from_binary_mask(box_mask_lvl.todense())
+        for region in regions:  # at the moment there is only one region
+            bbox_coordinates = region_coordinates(region)
+            yield from self._grid_coordinates_from_bbox_coordinates(
+                bbox_coordinates, slide
+            )
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:252]: @ernestoarbitrio any better? if you have any suggestions, will be more than welcome

Comment (ernestoarbitrio) [reply src/histolab/tiler.py:252]: not really ... let's move on for now

**Pull request merged**